File: Zigzag Conversion - Leetcode 6/Zigzag Conversion - Leetcode 6.java
public class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) return s;

        List<StringBuilder> rows = new ArrayList<>();
        for (int i = 0; i < Math.min(numRows, s.length()); i++) {
            rows.add(new StringBuilder());
        }

        int i = 0;
        boolean goingDown = false;

        for (char c : s.toCharArray()) {
            rows.get(i).append(c);
            if (i == 0 || i == numRows - 1) goingDown = !goingDown;
            i += goingDown ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder row : rows) {
            result.append(row);
        }
        return result.toString();
    }
}

File: Zigzag Conversion - Leetcode 6/Zigzag Conversion - Leetcode 6.cpp
#include <string>
#include <vector>

class Solution {
public:
    std::string convert(const std::string& s, int numRows) {
        if (numRows == 1) return s;

        std::vector<std::string> rows(std::min(numRows, int(s.length())));
        int i = 0;
        bool goingDown = false;

        for (char c : s) {
            rows[i] += c;
            if (i == 0 || i == numRows - 1) goingDown = !goingDown;
            i += goingDown ? 1 : -1;
        }

        std::string result;
        for (const std::string& row : rows) {
            result += row;
        }
        return result;
    }
};
